src/reanimated2/animation/spring.ts

```diff
@@ -93,14 +121,56 @@ export function withSpring(
         underDampedEnvelope *
           (cos1 * (v0 + zeta * omega0 * x0) - omega1 * x0 * sin1);
 
-      // critically damped
-      const criticallyDampedEnvelope = Math.exp(-omega0 * t);
-      const criticallyDampedPosition =
-        toValue - criticallyDampedEnvelope * (x0 + (v0 + omega0 * x0) * t);
+      return { position: underDampedPosition, velocity: underDampedVelocity };
+    }
 
-      const criticallyDampedVelocity =
-        criticallyDampedEnvelope *
-        (v0 * (t * omega0 - 1) + t * x0 * omega0 * omega0);
+    function spring(animation: InnerSpringAnimation, now: Timestamp): boolean {
+      const { toValue, firstTimestamp, lastTimestamp, current, velocity } =
+        animation;
+
+      const timeFromStart = now - firstTimestamp;
+      if (userConfig?.duration && timeFromStart > userConfig.duration) {
+        animation.current = toValue;
+
+        // clear lastTimestamp to avoid using stale value by the next spring animation that starts after this one
+        animation.lastTimestamp = 0;
+        animation.firstTimestamp = 0;
```

Comment:
do we need to reset `firstTimestamp` here too? The comment above only mentions `lastTimestamp`